### PR TITLE
feat: Resolve environment variables using Ref and --parameters

### LINKED ISSUE #8: Feature Request: Read env vars from template
It's kind of annoying to have to add environment variables to a separate `envVars.json` file as well as to `template.yaml` - especially the ones that can otherwise be defined just once across all functions in `Globals.Function.Environment`. This would probably require adding some kind of support for `!Sub` command, so that we can pluck env values out of the host environment and thereby use local development values without polluting the `template.yaml` file.

### PR DESCRIPTION
**Pull request by cjol**
This PR introduces a `--parameters` parameter which is used to resolve `!Ref` intrinsic functions within the template. Currently this is only used to resolve `!Ref`s inside a `Environment.Variables` section.

Possible breaking changes:
 - `!Ref MyVar` is no longer transformed to `"Ref MyVar" but instead to `{Ref: "MyVar"}`. I don't know if any other parts of the code rely on this value being a string.
 - Environment variables which are defined as static strings in the template are now passed to functions (I would consider this a bugfix rather than a breaking change!)

Closes #8

### CODE REVIEW

File: src/parseSAMTemplate/index.js [hunk 1]
```diff
@@ -1,8 +1,3 @@
-import yaml from 'js-yaml';
+import { yamlParse } from 'yaml-cfn';
```
Comment (duartemendes) [inline src/parseSAMTemplate/index.js]: Nice one, don't know how I didn't found this package earlier 👌

**Review by duartemendes — CHANGES_REQUESTED** [review]
Code looks super neat, tests are also great, thank you!

One change that I want to propose is to change the name and format of the new cli parameter.

I do like the way you did it but the goal here is to keep consistency with `sam` cli, maintaing compatibilitiy and thus allowing users to switch between our proxy and sam cli easier.

`--parameters key1=value1,key2=value2` would become `--parameter-overrides ParameterKey=key1,ParameterValue=value1 ParameterKey=key2,ParameterValue=value2`.

I do think `sam` way is uglier and slightly more complicated but I also think keeping consistency is really important here.

Let me know what you think please.

File: src/parseSAMTemplate/index.js (hunk 1 above)
Comment (cjol) [reply src/parseSAMTemplate/index.js]: Yeah it's handy! Arguably we don't need parseSamTemplate anymore. I don't know how you feel about just removing it?

**Comment by cjol** [comment]
I think changing to --parameter-overrides in the format you suggested makes sense. I thought about doing it that way first but then I realised SAM is already inconsistent with CloudFormation so either way we're inconsistent with something. But overall I agree we might as well.be consistent with one of them, and SAM is the primary use case here.

I'm not going to be around to make those changes for the next week or so. Feel free to leave the PR open for me to come back to, or take over and make the changes directly.

**Comment by duartemendes** [comment]
Actually, just now I realised that this feature doesn't reflect what `--parameter-overrides` does. So keeping consistency with the name is pointless. I do see a problem if later we want to tackle sam's `--parameter-overrides` feature, we would have to use another name or do a breaking change. So since this feature is about overriding environment variables, what about renaming it to `--env-vars-overrides`, given that we already use `--env-vars` for the file?

Or maybe something like `--ref-overrides`? I'm happy to make this change, just want an opinion first @cjol :)

Regarding the format, `sam` seems to support both ways `ParameterKey=key,ParameterValue=value` and `key=value`. So I'm happy with keeping as is and we could optional add support for the first option or just do it later when needed/requested.

File: src/parseSAMTemplate/index.js (hunk 1 above)
Comment (duartemendes) [reply src/parseSAMTemplate/index.js]: Yeah I think it makes sense to just remove it 👌

Comment (duartemendes) [reply src/parseSAMTemplate/index.js]: done!

**Comment by cjol** [comment]
I think `--ref-overrides` makes most sense as a suggestion. Customising the environment vars is (potentially) just one application of using !Ref like this, so I don't think `env-vars-*` is so good.

**Comment by duarte-airsiders** [comment]
Thank you @cjol, I'll try to wrap this up this week.

**Comment by duartemendes** [comment]
Thank you @cjol, I'll try to wrap this up this week.

**Review by duartemendes — APPROVED** [review]

**Comment by duartemendes** [comment]
@cjol thank you again for your work and sorry it took me some time to wrap this up. This has now been published as of version 0.0.15.

**Pull request merged**